File: 053. 最大子数组和.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        #只要前面的累加和是负数就扔掉、从当前数字重新开始算，否则就带着一起走，一路上随时记下出现过的最大值。
        result = pre = nums[0]
        for i in range(1,len(nums)):
            pre = max(nums[i],pre+nums[i])
            result=max(result,pre)
        return result
            

File: 053. 最大子数组和.go
package main

import "math"

func maxSubArray(nums []int) int {
	// 只要遇到负数，就把 sum 清零
	// 先维护结果再清零，否则要是 sum 全为负数 res 得不到维护就会返回无限小
	res := math.MinInt
	sum := 0
	for i := range len(nums) {
		sum += nums[i]
		res = max(res, sum)
		if sum < 0 {
			sum = 0
			continue
		}
	}
	return res
}
